pangeo-binder/values.yaml

```diff
@@ -198,7 +198,15 @@ dask-gateway:
       workerStartTimeout: 600
       worker:
         extraPodConfig:
-#          serviceAccountName: pangeo
+          serviceAccountName: pangeo
+          automountServiceAccountToken: true
+          securityContext:
+            fsGroup: 100
+# jupyter singleuser container has: 
+#          containers:
+#            securityContext:
+#              runAsGroup: 0
+#              runAsUser: 1000
```

Comment:
I commented this out for now b/c I was getting an error. Not sure how to set it, but jupyterhub singleuser containers get this security context, and I image we want the same for the image running on dask workers? Do we want `runAsGroup: 1000` as well? @consideRatio might be able to advise here

Reply:
I think I know too little to provide a concrete suggestion, but...

- jupyter/docker-stacks use containers starting in the root group to do some tricks from their start scripts and assumes some stuff in the base-notebook Dockerfile. I don't think we do such magic.
- a pod can have a security context, and a container can have it - the pod will be the default of the containers in the pod unless specified on the container
- jupyter/docker-stacks have been using group 100 rather than 1000
- runAsUser / runAsGroup override the defaults of the image, and is a way to make Kubernetes aware of how the container will startup, which makes it able to enforce policies. Without policies, manually specifying this as the Dockerfile's values is pointless I think. But, if there are policies is not our control, so I would suggest setting this to the default of the Dockerfile we want to startup.

Reply:
This is values to be pasaed to the dask-gateway helm chart, I'm not sure how they translate without looking in the dask-gateway templates.

This project reference the dask-gateway helm chart v0.6.1

https://github.com/pangeo-data/pangeo-binder/blob/da809585b68d50134ea0fd545a220ba186caaedd/pangeo-binder/requirements.yaml#L15-L17

I didn't see dask-gateway allowing this to be configured in their helm chart in 0.6.1 or master: https://github.com/dask/dask-gateway/search?q=securityContext&type=Code

Reply:
thanks for the info @consideRatio ! Yes, I ran into the following error trying to set the container values, and assumed I had the syntax or indentation wrong:
```
    HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Pod in version \"v1\" cannot be handled as a Pod: v1.Pod.Spec: v1.PodSpec.Containers: []v1.Container: decode slice: expect [ or n, but found {, error found in #10 byte of ...|ainers\": {\"securityC|..., bigger context ...|tomountServiceAccountToken\": true, \"containers\": {\"securityContext\": {\"runAsGroup\": 0, \"runAsUser\": |...","reason":"BadRequest","code":400}
```

Reply:
Oh, it is under extraPodConfig! Then, concrete suggestion:

Set a securityContext with runAsUser and runAsGroup explicitly for containers to match whatever the Dockerfile likes to run as, but not on pods, since they can get a sidecar container that wants to run in a different manner, and then this can cause issues.

We may want to set the dask-gateway helm charts extraContainerConfig for both scheduler and worker. See https://github.com/dask/dask-gateway/blob/b4c3d1899283e1bbc23e4a28f585a9439c69a188/resources/helm/dask-gateway/values.yaml#L76-L78

And, that should be valid kubernetes api specs for a container: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#container-v1-core

so:

```yaml
gateway:
  backend:
    scheduler:
      securityContext:
        runAsGroup: either 100 or 1000 I guess
        runAsUser: 1000
    worker:
      securityContext:
        runAsGroup: either 100 or 1000 I guess
        runAsUser: 1000
```

Note that dask-gateway 0.6.1 is quite different from master currently.

Reply:
@scottyhq note that the `fsGroup` on the pod security policy is another matter that relates to the filesystem ownership of mounted folders, not what user is running the container.